common/src/main/java/io/pravega/common/concurrent/ScheduledQueue.java

```diff
@@ -0,0 +1,347 @@
+     */
+    @Override
+    public boolean add(E e) {
+        return offer(e);
+    }
+
+    /**
+     * Inserts the specified element into this delay queue.
+     *
+     * @param e the element to add
+     * @return {@code true}
+     * @throws NullPointerException if the specified element is null
+     */
+    @Override
+    public boolean offer(E e) {
+        long seq = itemsAdded.incrementAndGet();
+        if (!e.isDelayed()) {
+            readyTasks.add(e);
+        } else {
+            delayedTasks.put(new FireTime(e.getScheduledTimeNanos(), seq), e);
+        }
+        // This is done unconditionally even if delayed because it could be the
+        // new shortest delay in which case some thread should wake up to
+        // re-schedule its sleep.
+        blocker.release();
+        return true;
+    }
+
+    /**
+     * Inserts the specified element into this delay queue. As the queue is
+     * unbounded this method will never block.
+     *
+     * @param e the element to add
+     * @param timeout This parameter is ignored as the method never blocks
+     * @param unit This parameter is ignored as the method never blocks
+     * @return {@code true}
+     * @throws NullPointerException {@inheritDoc}
+     */
+    @Override
+    public boolean offer(E e, long timeout, TimeUnit unit) {
+        return offer(e);
+    }
+    
+    /**
+     * Inserts the specified element into this delay queue. As the queue is
+     * unbounded this method will never block.
+     *
+     * @param e the element to add
+     * @throws NullPointerException {@inheritDoc}
+     */
+    @Override
+    public void put(E e) {
+        offer(e);
+    }
+
+    /**
+     * Retrieves, but does not remove, the head of this queue, or
+     * returns {@code null} if this queue is empty. 
+     * 
+     * Unlike {@code poll}, this method returns the elements who's scheduled time has not yet arrived.
+     *
+     * @return the head of this queue, or {@code null} if this queue is empty
+     */
+    @Override
+    public E peek() {
+        val ready = readyTasks.peek();
+        if (ready == null) {            
+            val result = delayedTasks.firstEntry();
+            if (result == null) {
+                return null;
+            } else {            
+                return result.getValue();
+            }
+        } else {
+            return ready;
+        }
+    }
+
+    /**
+     * Returns the size of this collection. 
+     * The value is only guaranteed to be accurate if there are not concurrent operations being performed.
+     * If there are, it may reflect the operation or not.
+     * This call is O(1).
+     */
+    @Override
+    public int size() {
+        return (int) Long.min(itemsAdded.get() - itemsRemoved.get(), Integer.MAX_VALUE);
+    }
+
+    /**
+     * Always returns {@code Integer.MAX_VALUE} because
+     * a {@code OptionalDealyQueue} is not capacity constrained.
```

Comment:
Where is this class? Are you referring to `ScheduledQueue`?

Reply:
Yes